libcxx/include/__functional/move_only_function_impl.h

```diff
@@ -0,0 +1,233 @@
+
+#ifdef _LIBCPP_ABI_MOVE_ONLY_FUNCTION_TRIVIAL_ABI
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI [[_Clang::__trivial_abi__]]
+#else
+#  define _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI
+#endif
+
+template <class...>
+class move_only_function;
+
+template <class _ReturnT, class... _ArgTypes>
+class _LIBCPP_MOVE_ONLY_FUNCTION_TRIVIAL_ABI move_only_function<_ReturnT(
+    _ArgTypes...) _LIBCPP_MOVE_ONLY_FUNCTION_CVREF noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT)> {
+private:
+  static constexpr size_t __buffer_size_      = 3 * sizeof(void*);
+  static constexpr size_t __buffer_alignment_ = alignof(void*);
+  using _BufferT                              = __small_buffer<__buffer_size_, __buffer_alignment_>;
+
+  using _TrivialVTable    = _MoveOnlyFunctionTrivialVTable<_BufferT, _ReturnT, _ArgTypes...>;
+  using _NonTrivialVTable = _MoveOnlyFunctionNonTrivialVTable<_BufferT, _ReturnT, _ArgTypes...>;
+
+  template <class _Functor>
+  static constexpr _TrivialVTable __trivial_vtable_ = {
+      .__call_ = [](_BufferT& __buffer, _ArgTypes... __args) noexcept(_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT) -> _ReturnT {
+        return std::invoke_r<_ReturnT>(
+            static_cast<_Functor _LIBCPP_MOVE_ONLY_FUNCTION_INVOKE_QUALS>(*__buffer.__get<_Functor>()),
+            std::forward<_ArgTypes>(__args)...);
+      }};
+
+  template <class _Functor>
+  static constexpr _NonTrivialVTable __non_trivial_vtable_{
+      __trivial_vtable_<_Functor>,
+      [](_BufferT& __buffer) noexcept -> void {
+        std::destroy_at(__buffer.__get<_Functor>());
+        __buffer.__dealloc<_Functor>();
+      },
+  };
+
+  template <class _Functor>
+  _LIBCPP_HIDE_FROM_ABI __pointer_bool_pair<const _TrivialVTable*> __get_vptr() {
+    if constexpr (_BufferT::__fits_in_buffer<_Functor> && is_trivially_destructible_v<_Functor>) {
+      return {&__trivial_vtable_<_Functor>, false};
+    } else {
+      return {&__non_trivial_vtable_<_Functor>, true};
+    }
+  }
+
+  template <class _VT>
+  static constexpr bool __is_callable_from = [] {
+    using _DVT = decay_t<_VT>;
+    if (_LIBCPP_MOVE_ONLY_FUNCTION_NOEXCEPT) {
+      return is_nothrow_invocable_r_v<_ReturnT, _DVT _LIBCPP_MOVE_ONLY_FUNCTION_CVREF, _ArgTypes...> &&
+             is_nothrow_invocable_r_v<_ReturnT, _DVT _LIBCPP_MOVE_ONLY_FUNCTION_INVOKE_QUALS, _ArgTypes...>;
+    } else {
+      return is_invocable_r_v<_ReturnT, _DVT _LIBCPP_MOVE_ONLY_FUNCTION_CVREF, _ArgTypes...> &&
+             is_invocable_r_v<_ReturnT, _DVT _LIBCPP_MOVE_ONLY_FUNCTION_INVOKE_QUALS, _ArgTypes...>;
+    }
+  }();
+
+  template <class _Func, class... _Args>
+  _LIBCPP_HIDE_FROM_ABI void __construct(_Args&&... __args) {
+    static_assert(is_constructible_v<decay_t<_Func>, _Func>);
+
+    using _StoredFunc = decay_t<_Func>;
+    __vtable_         = __get_vptr<_StoredFunc>();
+    __buffer_.__construct<_StoredFunc>(std::forward<_Args>(__args)...);
+  }
+
+  _LIBCPP_HIDE_FROM_ABI void __reset() {
+    if (__vtable_.__get_value())
```

Comment:
IIUC, I think this is the only place where the `bool` is used. IIUC, the only purpose of the bool is an optimisation that skips the virtual function call  for trivially destructible type, I think we can avoid using this bool at all. You can just have a vtable that contains "destroy" function pointer all the time. for trivially destrutible type, you just make the destroy function pointer nullptr. and here you simply check `__vtable_.destroy_ != nullptr`

Reply:
The `bool` is in the low bit of the pointer to avoid the need to even look at the vtable in the destructor. The cost of zeroing the lower bits of the pointer is pretty much non-existent on modern platforms (the delay is at most a single cycle - most likely not even that) and we gain that we avoid a potential cache miss during destruction.

Reply:
There's other cost here. 

There's complexity, there's ABI risk. To speak on the ABI risk.:

 From my investigations Clang has trouble optimizing around this, where it doesn't have trouble optimizing around what @huixie90 is suggesting, which I've implemented and tested out.

 And there's two reviewers asking you to remove it.

Reply:
Then provide the fucking evidence.

Reply:
So far the conclusion was that this `bool` is used to optimise the destructor for trivially destructible types, to save one vtable ptr indirection, at the cost of code complexity and some pointer pokery in other operations like `operator()`. And the cost is debatable. (At least, you two seem to disagree on this cost)

Personally, I'd suggest to create some benchmarks, focusing on some realistic use cases and see if optimisation focusing on the destructor is worth it.

Reply:
The main thing I'm objecting to currently is Eric's behaviour. He's showing zero evidence for his claims while being extremely patronizing to the point of being insulting [1] and claiming he's got knowledge of other's mental states which he obviously has no access to [2].

[1]
> @philnik777 I would encourage you to read about the importance of learning to "kill your darlings". The saying comes from creating writing, but is equally important here.

[2] (#94324)
> I think we discovered that this type is not needed (and somewhat harmful to the optimizer)


Anyways, yes, benchmarks would be nice, but I'm not convinced we can produce any micro-benchmarks that are close to realistic. Anything we produce will be highly biased based on the usage we expect. Cache misses on the scale of a program are also extremely hard to emulate in any micro benchmark. If you have any suggestions they're very welcome. I'd love to have some more detailed data on this.

Reply:
@philnik777 It seems we're having trouble communicating. I'm happy to meet with you over video chat, hopefully that can clear up any miscommunications.

I promise there is no malice or ill-intent. 

The comment about "killing your darlings" is  honest advice. Perhaps the phrase comes off poorly if you've never heard it before. Here's [a blog post](https://www.masterclass.com/articles/what-does-it-mean-to-kill-your-darlings#what-are-the-origins-of-the-phrase-kill-your-darlings) explaining the saying.

> What Does It Mean to Kill Your Darlings?
Writing is a painful process and most experienced writers will tell you that good writing involves substantial rewriting. An essential part of the rewriting process is combing through your work and cutting out material that isn’t essential. Sometimes this means we have to lose things that we are proud of and attached to. When you edit out material like this, you are killing your darlings.

I've provided a complete & test-passing implementation to compare against. It's the baseline I used for this review. I'm confident my results are repeatable. Please let me know  If can help you reproduce the output I've stated.

Reply:
> I've provided a complete & test-passing implementation to compare against. It's the baseline I used for this review. I'm confident my results are repeatable.

Sorry, I'm trying to follow the discussion after the fact but I can't find where you posted that implementation. Can you please share the link again?

Reply:
@ldionne You can find it in the [comment here](https://github.com/llvm/llvm-project/pull/94670/files#r1630389880).